## Part 1: Preprocessing


In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv(
    "https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv"
)
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [3]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[["Attrition", "Department"]]
y_df

Attrition              Department
0          Yes                   Sales
1           No  Research & Development
2          Yes  Research & Development
3           No  Research & Development
4           No  Research & Development
...        ...                     ...
1465        No  Research & Development
1466        No  Research & Development
1467        No  Research & Development
1468        No                   Sales
1469        No  Research & Development

[1470 rows x 2 columns]

In [4]:
# Create a list of at least 10 column names to use as X data
sel_col = [
    "Education",
    "Age",
    "DistanceFromHome",
    "JobSatisfaction",
    "OverTime",
    "StockOptionLevel",
    "WorkLifeBalance",
    "YearsAtCompany",
    "YearsSinceLastPromotion",
    "NumCompaniesWorked",
]

# Create X_df using your selected columns
X_df = attrition_df[sel_col]

# Show the data types for X_df
X_df.dtypes

Education                   int64
Age                         int64
DistanceFromHome            int64
JobSatisfaction             int64
OverTime                   object
StockOptionLevel            int64
WorkLifeBalance             int64
YearsAtCompany              int64
YearsSinceLastPromotion     int64
NumCompaniesWorked          int64
dtype: object

In [5]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
X_df["OverTime"].value_counts()

OverTime
No     1054
Yes     416
Name: count, dtype: int64

In [6]:
X_df["OverTime"] = X_df["OverTime"].map({"Yes": 1, "No": 0})
X_df["OverTime"] = X_df["OverTime"].astype("int64")
X_df["OverTime"].value_counts()

/var/folders/x2/qh793d497wd3pr7qgwn4zkmc0000gn/T/ipykernel_84831/395158491.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_df["OverTime"] = X_df["OverTime"].map({"Yes": 1, "No": 0})
/var/folders/x2/qh793d497wd3pr7qgwn4zkmc0000gn/T/ipykernel_84831/395158491.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_df["OverTime"] = X_df["OverTime"].astype("int64")


OverTime
0    1054
1     416
Name: count, dtype: int64

In [7]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=1)

In [8]:
# Create a StandardScaler
X_scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [9]:
from sklearn.preprocessing import OneHotEncoder

# Create a OneHotEncoder for the Department column
department_encoder = OneHotEncoder(sparse_output=False)
# Fit the encoder to the training data
y_department_train = department_encoder.fit_transform(y_train[["Department"]])
y_department_test = department_encoder.transform(y_test[["Department"]])
y_department_train
# Create two new variables by applying the encoder
# to the training and testing data

array([[0., 1., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       ...,
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 0., 1.]])

In [10]:
# Create a OneHotEncoder for the Attrition column
attrition_encoder = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
y_attrition_train = attrition_encoder.fit_transform(y_train[["Attrition"]])
y_attrition_test = attrition_encoder.transform(y_test[["Attrition"]])
# Create two new variables by applying the encoder
# to the training and testing data
y_attrition_train

array([[1., 0.],
       [1., 0.],
       [0., 1.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

## Create, Compile, and Train the Model


In [11]:
# Find the number of columns in the X training data
input_nodes = X_train.shape[1]

# Create the input layer
input_layer = layers.Input(shape=(input_nodes,), name="input_layer")

# Create at least two shared layers
shared_layer1 = layers.Dense(64, activation="relu", name="shared_layer1")(input_layer)
shared_layer2 = layers.Dense(128, activation="relu", name="shared_layer2")(
    shared_layer1
)

In [12]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
department_hidden = layers.Dense(32, activation="relu", name="department_hidden")(
    shared_layer2
)

# Create the output layer
department_output = layers.Dense(3, activation="softmax", name="department_output")(
    department_hidden
)

In [13]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden = layers.Dense(32, activation="relu", name="attrition_hidden")(
    shared_layer2
)
# Create the output layer
attrition_output = layers.Dense(2, activation="softmax", name="attrition_output")(
    attrition_hidden
)

In [14]:
# Create the model
model = Model(inputs=input_layer, outputs=[department_output, attrition_output])

# Compile the model
model.compile(
    optimizer="adam",
    loss={
        "department_output": "categorical_crossentropy",
        "attrition_output": "categorical_crossentropy",
    },
    metrics={
        "department_output": "accuracy",
        "attrition_output": "accuracy",
    },
)

# Summarize the model
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_layer (InputLayer)       [(None, 10)]         0           []                               
                                                                                                  
 shared_layer1 (Dense)          (None, 64)           704         ['input_layer[0][0]']            
                                                                                                  
 shared_layer2 (Dense)          (None, 128)          8320        ['shared_layer1[0][0]']          
                                                                                                  
 department_hidden (Dense)      (None, 32)           4128        ['shared_layer2[0][0]']          
                                                                                              

In [15]:
# Fit the model
model.fit(
    X_train_scaled,
    {"department_output": y_department_train, "attrition_output": y_attrition_train},
    epochs=50,
    batch_size=32,
    validation_split=0.2,
)

2024-04-24 15:01:14.999099: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz


Epoch 1/50
28/28 [==============================] - 0s 5ms/step - loss: 1.4305 - department_output_loss: 0.9201 - attrition_output_loss: 0.5104 - department_output_accuracy: 0.5800 - attrition_output_accuracy: 0.8104 - val_loss: 1.3218 - val_department_output_loss: 0.8306 - val_attrition_output_loss: 0.4912 - val_department_output_accuracy: 0.6335 - val_attrition_output_accuracy: 0.8281
Epoch 2/50
28/28 [==============================] - 0s 1ms/step - loss: 1.1629 - department_output_loss: 0.7702 - attrition_output_loss: 0.3927 - department_output_accuracy: 0.6640 - attrition_output_accuracy: 0.8524 - val_loss: 1.2985 - val_department_output_loss: 0.8352 - val_attrition_output_loss: 0.4632 - val_department_output_accuracy: 0.6335 - val_attrition_output_accuracy: 0.8235
Epoch 3/50
28/28 [==============================] - 0s 1ms/step - loss: 1.1156 - department_output_loss: 0.7594 - attrition_output_loss: 0.3562 - department_output_accuracy: 0.6640 - attrition_output_accuracy: 0.8558 - v

28/28 [==============================] - 0s 1ms/step - loss: 1.0266 - department_output_loss: 0.7290 - attrition_output_loss: 0.2976 - department_output_accuracy: 0.6663 - attrition_output_accuracy: 0.8831 - val_loss: 1.3330 - val_department_output_loss: 0.8621 - val_attrition_output_loss: 0.4709 - val_department_output_accuracy: 0.6199 - val_attrition_output_accuracy: 0.8281
Epoch 8/50
28/28 [==============================] - 0s 1ms/step - loss: 1.0099 - department_output_loss: 0.7207 - attrition_output_loss: 0.2892 - department_output_accuracy: 0.6686 - attrition_output_accuracy: 0.8876 - val_loss: 1.3597 - val_department_output_loss: 0.8787 - val_attrition_output_loss: 0.4811 - val_department_output_accuracy: 0.6154 - val_attrition_output_accuracy: 0.8235
Epoch 9/50
28/28 [==============================] - 0s 1ms/step - loss: 0.9910 - department_output_loss: 0.7099 - attrition_output_loss: 0.2811 - department_output_accuracy: 0.6708 - attrition_output_accuracy: 0.8944 - val_loss: 1.

In [17]:
# Evaluate the model with the testing data
model_eval = model.evaluate(
    X_test_scaled,
    {"department_output": y_department_test, "attrition_output": y_attrition_test},
)
model_eval

12/12 [==============================] - 0s 573us/step - loss: 2.3957 - department_output_loss: 1.4063 - attrition_output_loss: 0.9894 - department_output_accuracy: 0.5598 - attrition_output_accuracy: 0.8043


[2.3957135677337646,
 1.406286597251892,
 0.9894269108772278,
 0.5597826242446899,
 0.804347813129425]

In [18]:
# Print the accuracy for both department and attrition
# Print the quality and color accuracy
print(f"Department predictions accuracy: {model_eval[3]}")
print(f"Attrition predictions accuracy: {model_eval[4]}")

Department predictions accuracy: 0.5597826242446899
Attrition predictions accuracy: 0.804347813129425


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?


YOUR ANSWERS HERE

1. Accuracy isn't always the best way to judge our model. For 'Department' predictions where the accuracy is just about 56%, it hints that maybe the data isn't evenly spread across all classes. And for 'Attrition', even though we got 80% accuracy, it’s good to check other stats like precision or recall to really understand if we’re catching all the right cases, especially the tricky ones.
2. I went with the softmax activation function for both outputs. It’s great for cases where you need to pick one class out of many, like our departments, because it gives a probability for each class, making it clear which one’s the most likely pick.
3.

- **More Data**: Especially for departments where accuracy is low, more examples could help.
- **Tweak the Knobs**: Playing around with the model’s settings, like how fast it learns or how it processes data in batches, might boost performance.
- **Mix Up the Features**: Maybe adding some new features or tweaking the existing ones could help the model see things more clearly."
